Custom roles and permissions (#523)
powersync-ja/powersync-docs: tools/powersync-dashboard.mdx

    - **Instance**: Development
    - **Instance**: Production

## Dashboard Overview

The PowerSync Dashboard is organized into three main levels, each providing different functionality:

1. **Organization Level** - Manage projects, team members, organization and billing settings
2. **Account Level** - Manage your personal account settings and access tokens
3. **Project & Instance Level** - Configure and monitor your PowerSync instances

## Organization Level

URL structure: `https://dashboard.powersync.com/org/{orgId}/projects`

At the organization level, you can manage your PowerSync projects and organization-wide settings. Navigate to your organization to access:

- **Projects** - View all projects in your organization, create new projects, and manage project settings
- **Team** - Invite team members to your organization, manage their access with [roles and permissions](#roles-and-permissions), and remove access
- **Plans & Billing** - View or update your PowerSync subscription plan and manage billing details
- **Plan Usage** - Monitor usage metrics across all projects in your organization for the current billing cycle
- **Settings** - Update the organization name

> ## Roles and Permissions

Reviewer: Maybe this section should be moved to the end of this page and not placed in-between `Organization level` and `Account level` (feels a but confusing)

> A role is a named set of permissions that controls what a member can see and do in your organization. You assign a role when inviting a member, and you can change a member's roles at any time. A member can hold more than one role. Manage roles from the **Roles** tab on the **Team** page.

Reviewer 2: A part of this is too loosely written for users below the Team plan, e.g. they can't "Manage roles from the **Roles** tab on the **Team** page." - for them it's just a reference.

<Frame caption="The Roles tab on the Team page, listing built-in and custom roles with their grants and assigned members">
  ![](/images/usage/tools/dashboard-team-roles.png)
</Frame>

> ### Built-in Roles

Reviewer: ```suggestion
### System Roles
```

Every organization includes built-in roles that you can assign but cannot edit or delete:

- **Owner** - Full access to the organization, its projects, instances, team, and billing.

> - **Developer** - A broad working set of permissions for building and operating projects and instances.

Reviewer: > A broad working set of permissions for building and operating projects and instances.

This does not match what is currently being shown for `Developer` role. Should we show this description instead of `Perform operations necessary for developing on an app` in the Dashboard?

Author: I've changed it to match the role description in the dashboard

Reviewer: Updated it again with the latest description

You can see the exact grants for any role in the Roles tab. To grant access that differs from the built-in roles, create a custom role.

### Custom Roles

Custom roles are available on the [Team and Enterprise plans](https://www.powersync.com/pricing).

A custom role combines a set of permissions with a project scope. Each permission covers one area of the dashboard and has one or more access levels. A higher level includes everything the lower level allows. A role needs at least one permission.

The access levels mean the same thing across every permission:

- **View** gives read-only access to that area. The member can see it but cannot change anything.
- **Deploy** applies only to Instances. It gives View access plus the ability to deploy changes to an instance, such as updating the sync config and database connections. It does not allow creating or deleting instances.

> - **Manage** gives full access to that area, including creating, editing, and deleting.

Reviewer 2: ```suggestion
- **Manage** gives full access to instances, including creating, editing, deploying and deleting.
```

Author: I added the deploying part but left it generic because Manage applies to more permissions than just instances

The table below lists each permission, its available access levels, and what those levels grant.

| Permission | Access levels | What it grants |
| --- | --- | --- |
| Projects | View, Manage | View projects. Manage adds creating, renaming, and deleting projects. |
| Instances | View, Deploy, Manage | View instances. Deploy adds deploying changes such as sync config and database connections. Manage adds creating and deleting instances. |
| Instance logs | View | View instance logs. |
| Alert rules | View, Manage | View alert rules. Manage adds creating, editing, and deleting them. |

> | Notification configurations | View, Manage | View notification configurations. Manage adds creating, editing, deleting, and testing them. |

Reviewer 2: In the dashboard we call these "Notification Rules" - we should match that wording. @Manrich121 the same issue exists in the dashboard code - we should update there.

| Private Endpoints | View, Manage | View Private Endpoints. Manage adds creating, editing, and deleting them. |
| Team | View, Manage | View members. Manage adds inviting and removing members and changing their role assignments. |
| Billing | View, Manage | View billing details and the current plan. Manage adds updating billing and changing the plan. |

> | Organization settings | Manage | Update organization configuration, such as the organization name. |

Reviewer 2: I think here we need to explain why "View" is not a specific access level

Some permissions depend on others, because access to the parent area is required to reach the child. Instances, Alert rules, and Notification configurations require at least View on Projects, and Instance logs requires at least View on Instances. The builder adds the required parent permissions for you when you select a dependent one.

<Frame caption="The custom role builder, showing permission rows with access levels and the project scope picker">
  ![](/images/usage/tools/dashboard-role-builder.png)
</Frame>

### Project Scope

> A custom role applies to either all projects in the organization or to a specific set of projects that you choose. Scope affects the permissions that operate at the project level: Projects, Instances, Instance logs, Alert rules, and Notification configurations. The remaining permissions (Private Endpoints, Team, Billing, and Organization settings) always apply to the whole organization.

Reviewer: The paragraph is difficult to read. A few comments:
1. Not all Custom roles are scoped to projects (all or selected), for example; a role can have only Billing permissions, which is org scoped.
2. If this paragraph is trying to convey what scope is, it would be better to list it, org-, project- and instance

> A project-scoped member is granted their permissions inside the selected projects and has no access outside them. The Manage level on Projects is only available on roles scoped to all projects, since creating and deleting projects is an organization-wide action.

Reviewer: ```suggestion
A project-scoped member is granted their permissions inside the selected projects and has no access outside them. The Manage level on Projects is only available on roles scoped to all projects, since creating and deleting projects is an organization-scoped action.
```

### How Permissions Appear in the Dashboard

> Members only see and do what their roles allow. Where a member lacks permission for an action, the related control stays visible but disabled, so it is clear which capabilities an administrator could grant. Opening a page the member cannot view shows an access message rather than the page content. The dashboard reflects permissions as a convenience. Access is always enforced by the PowerSync backend.

Reviewer: ```suggestion
Members only has access to and allow actions that their roles allow. Where a member lacks permission for an action, the related control stays visible but disabled. Opening a page the member cannot view shows an "You do not have permission" message rather than the page content. The dashboard reflects permissions as a convenience. Access is always enforced by the PowerSync backend.
```

Versions are specified as `major.minor.patch`. When locked, only new `.patch` releases will automatically be applied to the instance.

**Downgrade limitations:** Not all downgrade paths are available automatically. If you need to downgrade to an older version, please [contact our team](/resources/contact-us) for assistance.

## Roles and Permissions

A role is a named set of permissions that controls what a member can see and do in your organization. You assign one or more roles to a member when you invite them, and you can change a member's roles at any time. The **Roles** tab on the **Team** page lists the roles available in your organization. System roles are available on every plan, and creating custom roles requires the [Team or Enterprise plan](https://www.powersync.com/pricing).

<Frame caption="The Roles tab on the Team page, listing system and custom roles with their grants and assigned members">
  ![](/images/usage/tools/dashboard-team-roles.png)
</Frame>

### System Roles

Every organization includes system roles that you can assign but cannot edit or delete:

- **Owner** - Full access to the organization, its projects, instances, team, and billing.

> - **Developer** - Perform operations necessary for developing on an app.

Reviewer 2: See feedback to this wording here https://docs.google.com/document/d/12rHnQt8bUqfiatb8t2FDMdZmyknl8QSXYEmDBXMq8oU/edit?tab=t.0#heading=h.v0ifqun2c8bs

> You can see the exact grants for any role in the Roles tab. To grant access that differs from the system roles, create a custom role.

Reviewer 2: `You can see the exact grants for any role in the Roles tab` Is this true for the system roles? For those my experience is you can just see the high level description.

Reviewer 2: Feels like we could remove this sentence

### Custom Roles

Custom roles are available on the [Team and Enterprise plans](https://www.powersync.com/pricing).

> A custom role is a set of permissions that you choose. Each permission covers one area of the dashboard and has one or more access levels. A higher level includes everything the lower level allows. A role needs at least one permission.

Reviewer 2: `covers one area of the dashboard` - I don't think we should limit it to the dashboard, since roles can also be assigned to PATs. We can say something like "covers one operational area", but honestly I feel that's kinda intuitive

| Private Endpoints | View, Manage | View Private Endpoints. Manage adds creating, editing, and deleting them. |
| Team | View, Manage | View members. Manage adds inviting and removing members and changing their role assignments. |
| Billing | View, Manage | View billing details and the current plan. Manage adds updating billing and changing the plan. |
| Organization settings | Manage | Update organization configuration, such as the organization name. |

Not every permission offers all access levels. Each area shows only the levels that apply to it: Instance logs are read-only (View only), and Organization settings have no separate read-only view (Manage only).

Some permissions depend on others, because access to the parent area is required to reach the child. Instances, Alert Rules, and Notification Rules require at least View on Projects, and Instance logs requires at least View on Instances. The builder adds the required parent permissions for you when you select a dependent one.

<Frame caption="The custom role builder, showing permission rows with access levels and the project scope picker">
  ![](/images/usage/tools/dashboard-role-builder.png)
</Frame>

### Project Scope

> Permissions fall into two groups. Organization-level permissions (Private Endpoints, Team, Billing, and Organization settings) always apply across the whole organization. Project-level permissions (Projects, Instances, Instance logs, Alert Rules, and Notification Rules) can be scoped to all projects or to a specific set of projects that you choose.

Reviewer 2: Aren't there 3 groups? Organization, Project, Instance? Project-level permissions just being quite minimal. From the UI  breakdown we have these 3 groups (e.g. in the permissions dropdown) and it aligns nicely with the high level hierarchy https://docs.powersync.com/tools/powersync-dashboard#hierarchy-organization-project-instance

Reviewer: Agree, update this now.

Project scope only applies to a role that includes project-level permissions. A role made up only of organization-level permissions, such as a Billing-only role, applies across the organization regardless of scope.

A project-scoped member is granted their permissions inside the selected projects and has no access outside them. The Manage level on Projects is only available on roles scoped to all projects, since creating and deleting projects is an organization-scoped action.

> ### How Permissions Appear in the Dashboard

Reviewer 2: Do we really need this? Feels a bit like AI generated info overload. Feels intuitive enough that if a user doesn't have permission they will in some way be blocked - I also don't think we should say something like "will stay visible but disabled" explicitly, because we might not be consistent with that or choose a different means to block them for some cases?

Reviewer: Removed
